File: tests/factory/interop.ts
const fs = require('node:fs');
const path = require('node:path');

type J = any;
const root: string = process.env.FACTORY_INTEROP_ROOT || path.resolve(__dirname, '../..');
const load = (rel: string): J => JSON.parse(fs.readFileSync(path.join(root, rel), 'utf8'));
const assert = (ok: boolean, msg: string): void => { if (!ok) throw new Error(msg); };
const canonical = /^[a-z][a-z0-9-]*:[0-7][0-9A-HJKMNP-TV-Z]{25}$/;
const kindRef = (kind: string, value: unknown): boolean => typeof value === 'string' && new RegExp(`^${kind}:[0-7][0-9A-HJKMNP-TV-Z]{25}$`).test(value);
const subjectKey = (s: J): string => `${s.subjectRef}|${s.subjectRevision}|${s.stateRef}`;

function checkOwners(node: J, where = '#'): void {
  if (!node || typeof node !== 'object' || Array.isArray(node)) return;
  if (node.properties) for (const [name, raw] of Object.entries(node.properties as Record<string, J>)) {
    const child: J = raw;
    assert(typeof child['x-semantic-owner'] === 'string' && child['x-semantic-owner'].trim().length > 0, `${where}/${name} missing one semantic owner`);
    checkOwners(child, `${where}/${name}`);
  }
  if (node.$defs) for (const [name, child] of Object.entries(node.$defs as Record<string, J>)) checkOwners(child, `${where}/$defs/${name}`);
  if (node.items && typeof node.items === 'object') checkOwners(node.items, `${where}/items`);
  if (node.anyOf) (node.anyOf as J[]).forEach((child, i) => checkOwners(child, `${where}/anyOf/${i}`));
}

function rejectsAnti(item: J): boolean {
  const v = item.value || {};
  switch (item.id) {
    case 'action-as-capability-identity-collapse': return kindRef('action', v.actionRef) && !kindRef('capability', v.capabilityRef);
    case 'binding-as-ref': return !!v.binding && Object.prototype.hasOwnProperty.call(v.binding, 'bindingRef') && canonical.test(v.binding.bindingRef);
    case 'model-as-agent':
    case 'session-as-agent': return !kindRef('agent', v.agentRef);
    case 'provider-as-project': return !kindRef('project', v.projectRef);
    case 'stale-subject-state-evidence': return subjectKey(v.currentSubjectState) !== subjectKey(v.evidenceSubjectState);
    default: return false;
  }
}

const schemaSet: J = load('contracts/factory/interop/schema-set.json');
const fixtureSet: J = load('contracts/factory/fixtures/interop/fixture-set.json');
assert(schemaSet.schemaSetVersion === 'factory.interop/v1', 'schema set version');
assert(fixtureSet.fixtureSetVersion === 'factory.interop-fixtures/v1' && fixtureSet.contractVersion === 'factory.interop/v1', 'fixture versions');
const schemas = new Map<string, J>(schemaSet.schemas.map((entry: J) => [entry.instanceKey, entry]));
const sections = new Map<string, J>(fixtureSet.sections.map((entry: J) => [entry.instanceKey, entry]));
assert(schemas.size === sections.size && [...schemas.keys()].every((key) => sections.has(key)), 'schema/fixture manifest drift');
for (const [key, entry] of schemas) {
  const schema: J = load(entry.path), fixture: J = load(sections.get(key).path);
  assert(schema.$id === entry.id, `${key} schema id drift`); checkOwners(schema);
  const encoded = JSON.stringify(fixture); assert(JSON.stringify(JSON.parse(encoded)) === encoded, `${key} round-trip drift`);
  assert(fixture.contractVersion === 'factory.interop/v1', `${key} contract version`);
}

const identity = load(sections.get('identityCore').path), run = load(sections.get('runProjection').path), execution = load(sections.get('execution').path);
const action = load(sections.get('action').path), capability = load(sections.get('capability').path), context = load(sections.get('context').path);
const evidence = load(sections.get('evidenceAssessment').path), closure = load(sections.get('closureGate').path), demand = load(sections.get('demand').path);
const offer = load(sections.get('workcellOffer').path), binding = load(sections.get('binding').path), world = load(sections.get('materialisedWorld').path), ql = load(sections.get('ql').path);

assert(canonical.test(identity.identityEnvelope.ref) && !identity.identityEnvelope.ref.startsWith('factory:'), 'canonical identity grammar');
assert(identity.identityEnvelope.revision === identity.identityEnvelope.subjectState.subjectRevision && identity.identityEnvelope.ref === identity.identityEnvelope.subjectState.subjectRef, 'identity subject/revision drift');
assert(kindRef('project', identity.projectBinding.projectRef) && !String(identity.projectBinding.sourceRef).startsWith('project:'), 'ProjectBinding/source boundary');
assert(kindRef('run', run.runEnvelope.runRef) && run.runEnvelope.runMapAddress === `${run.runEnvelope.runRef}/map`, 'Run/RunMap address');
assert(run.runEnvelope.projectRef === identity.projectBinding.projectRef, 'Run ProjectRef');
assert(run.projectionProvenance.length >= 2 && new Set(run.projectionProvenance.map((p: J) => p.providerRef)).size >= 2 && new Set(run.projectionProvenance.map((p: J) => p.externalId)).size >= 2, 'projection change not exercised');
assert(run.projectionProvenance.every((p: J) => p.canonicalRef === identity.projectBinding.projectRef && p.projectedRevision === run.runEnvelope.runRevision), 'projection identity/revision drift');
const ex = execution.executionEnvelope;
assert(kindRef('execution', ex.executionRef) && kindRef('agent', ex.agentRef) && kindRef('agency', ex.agencyRef) && kindRef('agent-session', ex.agentSessionRef), 'execution actor refs');
assert(ex.runRef === run.runEnvelope.runRef && !ex.agentRef.startsWith('model:') && !ex.agentRef.startsWith('agent-session:'), 'execution identity collapse/drift');
assert(kindRef('action', action.actionDescriptor.actionRef) && kindRef('capability', capability.capabilityDescriptor.capabilityRef) && action.actionDescriptor.actionRef !== capability.capabilityDescriptor.capabilityRef, 'Action/Capability distinction');
assert(action.actionDescriptor.ownerProjectRef === identity.projectBinding.projectRef && ex.capabilityRefs.includes(capability.capabilityDescriptor.capabilityRef), 'Action/Capability ownership/projection drift');
const cx = context.contextResolution, available = new Set<string>(cx.availableRefs), retrieved = new Set<string>(cx.retrievedRefs), loaded = new Set<string>(cx.loadedRefs);
assert([...loaded].every((r) => retrieved.has(r)) && [...retrieved].every((r) => available.has(r)), 'Context availability/retrieval/load drift');
assert(cx.projectRef === identity.projectBinding.projectRef && ex.contextResolutionRef === cx.contextResolutionRef && ex.generationRef === cx.generationRef, 'Context/generation identity drift');
assert(context.generationProvenance.actionRefs.includes(action.actionDescriptor.actionRef) && context.generationProvenance.capabilityRefs.includes(capability.capabilityDescriptor.capabilityRef), 'Generation content drift');
const current = subjectKey(identity.identityEnvelope.subjectState);
[evidence.evidenceEnvelope.subjectState, evidence.assessmentEnvelope.subjectState, closure.closureEnvelope.subjectState, closure.gateDecisionEnvelope.subjectState].forEach((s: J) => assert(subjectKey(s) === current, 'verification subject-state mismatch'));
assert(evidence.assessmentEnvelope.evidenceRefs.includes(evidence.evidenceEnvelope.evidenceRef) && closure.closureEnvelope.evidenceRefs.includes(evidence.evidenceEnvelope.evidenceRef) && closure.closureEnvelope.assessmentRefs.includes(evidence.assessmentEnvelope.assessmentRef), 'verification relation drift');
assert(closure.gateDecisionEnvelope.closureRef === closure.closureEnvelope.closureRef, 'Gate/Closure drift');
assert(evidence.assessmentEnvelope.producerExecutionRef !== evidence.evidenceEnvelope.producerExecutionRef && evidence.assessmentEnvelope.independentFromExecutionRefs.includes(evidence.evidenceEnvelope.producerExecutionRef), 'independent assessment lineage');
assert(!Object.keys(demand.executionDemand).some((key) => key.toLowerCase().includes('provider')), 'provider leaked into ExecutionDemand');
assert(demand.executionDemand.projectRef === identity.projectBinding.projectRef && demand.executionDemand.runRef === run.runEnvelope.runRef, 'ExecutionDemand identity drift');
assert(demand.candidateMaterialisationDemand.executionDemandRef === demand.executionDemand.demandRef && demand.candidateMaterialisationDemand.candidateRef === identity.identityEnvelope.ref, 'Candidate materialisation demand drift');
assert(demand.executionDemand.requiredAffordances.every((a: string) => offer.workcellOffer.affordances.includes(a)), 'required affordance unavailable');
assert(binding.binding.workcellRef === offer.workcellOffer.workcellRef && !('ref' in binding.binding) && !('bindingRef' in binding.binding) && !canonical.test(binding.binding.bindingKey), 'Binding identity boundary');
assert(world.materialisedExecutionWorld.executionDemandRef === demand.executionDemand.demandRef && world.materialisedExecutionWorld.candidateRef === identity.identityEnvelope.ref && world.materialisedExecutionWorld.workcellRef === offer.workcellOffer.workcellRef && world.materialisedExecutionWorld.bindingKeys.includes(binding.binding.bindingKey), 'Materialised world relation drift');
assert(canonical.test(ql.qlComposition.targetRef), 'QL target Ref'); ['qlFormRef','qlAddress','lensRef','qlTarget'].forEach((key) => assert(typeof ql.qlComposition[key] === 'string' && ql.qlComposition[key].length > 0, `${key} missing`));

const anti = load(fixtureSet.antiFixturesPath); const expectedAnti = new Set(['action-as-capability-identity-collapse','binding-as-ref','model-as-agent','session-as-agent','provider-as-project','stale-subject-state-evidence']);
assert(anti.fixtureVersion === 'factory.interop-fixtures/v1' && anti.antiFixtures.length === expectedAnti.size && anti.antiFixtures.every((item: J) => expectedAnti.has(item.id)), 'anti-fixture set incomplete');
anti.antiFixtures.forEach((item: J) => assert(item.mustReject === true && rejectsAnti(item), `anti fixture unexpectedly accepted: ${item.id}`));
console.log(`TypeScript CR-001 interop PASS (${schemas.size} schema sections, ${anti.antiFixtures.length} anti-fixtures)`);

File: contracts/factory/reference/interop.py
"""Language-neutral CR-001 semantic conformance reference.

This module validates the shared JSON fixture corpus without becoming a client
runtime package. Node, TypeScript and Rust consumers implement the same rules
from the versioned schema and fixtures.
"""
from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Any, Mapping

CANONICAL_REF = re.compile(r"^factory:[a-z][a-z0-9-]*:[A-Za-z0-9][A-Za-z0-9._-]*$")
KIND_REF = lambda kind: re.compile(rf"^factory:{re.escape(kind)}:[A-Za-z0-9][A-Za-z0-9._-]*$")
QL_COORDINATE_KEYS = ("qlFormRef", "qlAddress", "lensRef", "sublensRef")


class InteropError(ValueError):
    pass


def _require(condition: bool, message: str) -> None:
    if not condition:
        raise InteropError(message)


def _subject_tuple(value: Mapping[str, Any]) -> tuple[str, str, int]:
    return (value.get("subjectRef"), value.get("stateRef"), value.get("revision"))


def _validate_ref(value: Any, kind: str | None = None) -> None:
    pattern = CANONICAL_REF if kind is None else KIND_REF(kind)
    _require(isinstance(value, str) and pattern.fullmatch(value) is not None, f"invalid {kind or 'canonical'} Ref: {value!r}")


def schema_owner_failures(schema: Mapping[str, Any]) -> list[str]:
    """Every declared Factory instance field in the schema has exactly one semantic owner."""
    failures: list[str] = []

    def walk(node: Any, path: str) -> None:
        if not isinstance(node, dict):
            return
        properties = node.get("properties")
        if isinstance(properties, dict):
            for name, child in properties.items():
                owner = child.get("x-semantic-owner") if isinstance(child, dict) else None
                if not isinstance(owner, str) or not owner.strip():
                    failures.append(f"{path}/{name} has no semantic owner")
                walk(child, f"{path}/{name}")
        defs = node.get("$defs")
        if isinstance(defs, dict):
            for name, child in defs.items():
                walk(child, f"{path}/$defs/{name}")
        items = node.get("items")
        if isinstance(items, dict):
            walk(items, f"{path}/items")

    walk(schema, "#")
    return failures


def _validate_ql_composition(
    ql: Mapping[str, Any], schema: Mapping[str, Any], ql_schema: Mapping[str, Any]
) -> None:
    parent = schema.get("properties", {}).get("qlComposition", {})
    _require(parent.get("$ref") == ql_schema.get("$id"), "parent QL schema reference drift")
    _require(parent.get("x-semantic-owner") == "Standalone QL/MEF module", "QL semantic owner drift")
    _require("qlComposition" not in schema.get("$defs", {}), "Factory must not duplicate QL composition definition")

    properties = ql_schema.get("properties")
    _require(isinstance(properties, dict), "QL schema properties missing")
    _require(set(ql).issubset(properties), "QL composition contains fields outside the standalone schema")
    _require(isinstance(ql.get("targetRef"), str) and bool(ql["targetRef"]), "QL targetRef is required")
    _require(any(key in ql for key in QL_COORDINATE_KEYS), "QL composition requires at least one coordinate")

    for key in QL_COORDINATE_KEYS:
        if key not in ql:
            continue
        pattern = properties.get(key, {}).get("pattern")
        _require(isinstance(pattern, str), f"QL schema pattern missing for {key}")
        value = ql[key]
        _require(isinstance(value, str) and re.fullmatch(pattern, value) is not None, f"invalid {key}: {value!r}")

    _validate_ref(ql["targetRef"])


def validate_contract(
    contract: Mapping[str, Any], schema: Mapping[str, Any], ql_schema: Mapping[str, Any]
) -> None:
    _require(schema.get("$schema") == "https://json-schema.org/draft/2020-12/schema", "interop schema must use JSON Schema 2020-12")
    owner_failures = schema_owner_failures(schema)
    _require(not owner_failures, "; ".join(owner_failures))
    required = schema.get("required", [])
    _require(set(contract) == set(required), "contract fields must exactly match the versioned schema surface")
    _require(contract.get("contractVersion") == "factory.interop/v1", "unsupported interop version")

    identity = contract["identityEnvelope"]
    _validate_ref(identity["ref"])
    _require(isinstance(identity["revision"], int) and identity["revision"] >= 1, "identity revision must be positive")
    current_state = identity["subjectState"]
    _validate_ref(current_state["subjectRef"])
    _require(current_state["subjectRef"] == identity["ref"], "identity subject-state must describe the same subject")
    _require(current_state["revision"] == identity["revision"], "identity revision and subject-state revision drift")

    action = contract["actionDescriptor"]
    capability = contract["capabilityDescriptor"]
    _validate_ref(action["actionRef"], "action")
    _validate_ref(capability["capabilityRef"], "capability")
    _validate_ref(action["ownerProjectRef"], "project")
    _require(action["actionRef"] != capability["capabilityRef"], "Action and Capability identity collapsed")

    project_binding = contract["projectBinding"]
    _validate_ref(project_binding["projectBindingRef"], "project-binding")
    _validate_ref(project_binding["projectRef"], "project")
    _require(project_binding["projectRef"] == action["ownerProjectRef"], "project ownership drift across Action and ProjectBinding")
    _require(not str(project_binding["sourceRef"]).startswith("factory:project:"), "source reference cannot replace Factory Project identity")

    run = contract["runEnvelope"]
    _validate_ref(run["runRef"], "run")
    _validate_ref(run["projectRef"], "project")
    _validate_ref(run["runMapRef"], "run-map")
    _require(run["projectRef"] == project_binding["projectRef"], "Run ProjectRef drift")

    execution = contract["executionEnvelope"]
    _validate_ref(execution["executionRef"], "execution")
    _validate_ref(execution["runRef"], "run")
    _validate_ref(execution["agentRef"], "agent")
    _validate_ref(execution["agencyRef"], "agency")
    _validate_ref(execution["agentSessionRef"], "agent-session")
    _require(execution["runRef"] == run["runRef"], "Execution RunRef drift")
    _require(not execution["agentRef"].startswith("factory:agent-session:"), "Agent cannot be an AgentSession")
    _require(not execution["agentRef"].startswith("model:"), "Agent cannot be a model")

    evidence = contract["evidenceEnvelope"]
    assessment = contract["assessmentEnvelope"]
    closure = contract["closureEnvelope"]
    gate = contract["gateDecisionEnvelope"]
    _validate_ref(evidence["evidenceRef"], "evidence")
    _validate_ref(assessment["assessmentRef"], "assessment")
    _validate_ref(closure["closureRef"], "closure")
    _validate_ref(gate["gateDecisionRef"], "gate-decision")
    _validate_ref(evidence["producerExecutionRef"], "execution")
    _validate_ref(assessment["producerExecutionRef"], "execution")
    for label, value in (("evidence", evidence), ("assessment", assessment), ("closure", closure), ("gate", gate)):
        _require(_subject_tuple(value["subjectState"]) == _subject_tuple(current_state), f"{label} subject-state is stale or mismatched")
    _require(evidence["evidenceRef"] in assessment["evidenceRefs"], "Assessment lost EvidenceRef")
    _require(evidence["evidenceRef"] in closure["evidenceRefs"], "Closure lost EvidenceRef")
    _require(assessment["assessmentRef"] in closure["assessmentRefs"], "Closure lost AssessmentRef")
    _require(closure["closureRef"] == gate["closureRef"], "Gate/Closure relation drift")
    _require(assessment["producerExecutionRef"] != evidence["producerExecutionRef"], "independent Assessment reused producer execution")
    _require(evidence["producerExecutionRef"] in assessment["independentFromExecutionRefs"], "Assessment independence lineage missing")

    demand = contract["executionDemand"]
    _validate_ref(demand["demandRef"], "execution-demand")
    _validate_ref(demand["projectRef"], "project")
    _validate_ref(demand["runRef"], "run")
    if demand["candidateRef"] is not None:
        _validate_ref(demand["candidateRef"], "candidate")
    _require(not any("provider" in key.lower() for key in demand), "ExecutionDemand leaked provider selection")
    _require(demand["projectRef"] == project_binding["projectRef"] and demand["runRef"] == run["runRef"], "ExecutionDemand semantic identity drift")
    _require(demand["candidateRef"] == identity["ref"], "ExecutionDemand CandidateRef drift")

    offer = contract["workcellOffer"]
    missing = set(demand["requiredAffordances"]) - set(offer["affordances"])
    _require(not missing, f"WorkcellOffer misses required affordances: {sorted(missing)}")

    binding = contract["binding"]
    _require(isinstance(binding["bindingId"], str) and binding["bindingId"].startswith("binding:"), "Binding must use Workcell binding identity")
    _require(not binding["bindingId"].startswith("factory:"), "Binding cannot masquerade as canonical Ref")
    _validate_ref(binding["executionRef"], "execution")
    _require(binding["executionRef"] == execution["executionRef"], "Binding execution relation drift")
    _require(binding["workcellRef"] == offer["workcellRef"], "Binding WorkcellRef drift")

    generation = contract["generationProvenance"]
    _validate_ref(generation["generationRef"], "generation")
    _validate_ref(generation["contextResolutionRef"], "context-resolution")
    _validate_ref(generation["projectRef"], "project")
    _require(generation["projectRef"] == project_binding["projectRef"], "Generation ProjectRef drift")
    _require(action["actionRef"] in generation["actionRefs"], "Generation lost ActionRef")
    _require(capability["capabilityRef"] in generation["capabilityRefs"], "Generation lost CapabilityRef")
    _require(execution["generationRef"] == generation["generationRef"], "Execution GenerationRef drift")

    projections = contract["projectionProvenance"]
    _require(isinstance(projections, list) and len(projections) >= 2, "at least two projections are required for provider-change fixture")
    providers = set()
    external_ids = set()
    for projection in projections:
        _validate_ref(projection["projectionRef"], "projection")
        _validate_ref(projection["canonicalRef"])
        _require(projection["canonicalRef"] == project_binding["projectRef"], "projection lost canonical identity")
        _require(projection["projectedRevision"] >= 1, "projection revision invalid")
        _require(projection["generationRef"] == generation["generationRef"], "projection GenerationRef drift")
        providers.add(projection["providerRef"])
        external_ids.add(projection["externalId"])
    _require(len(providers) >= 2 and len(external_ids) >= 2, "provider/projection-change fixture must actually change provider and external identity")

    _validate_ql_composition(contract["qlComposition"], schema, ql_schema)


def reject_anti_fixture(item: Mapping[str, Any]) -> None:
    fixture_id = item.get("id")
    value = item.get("value", {})
    if fixture_id == "action-as-capability-identity-collapse":
        _validate_ref(value.get("actionRef"), "action")
        _validate_ref(value.get("capabilityRef"), "capability")
        _require(value["actionRef"] != value["capabilityRef"], "Action and Capability identity collapsed")
        return
    if fixture_id == "binding-as-ref":
        _require(isinstance(value.get("bindingId"), str) and value["bindingId"].startswith("binding:"), "Binding cannot be a canonical Ref")
        return
    if fixture_id in {"model-as-agent", "session-as-agent"}:
        _validate_ref(value.get("agentRef"), "agent")
        return
    if fixture_id == "provider-as-project":
        _validate_ref(value.get("projectRef"), "project")
        return
    if fixture_id == "stale-subject-state-evidence":
        _require(_subject_tuple(value["currentSubjectState"]) == _subject_tuple(value["evidenceSubjectState"]), "stale or mismatched Evidence cannot be current")
        return
    raise InteropError(f"unknown anti-fixture: {fixture_id}")


def validate_fixture_document(
    document: Mapping[str, Any], schema: Mapping[str, Any], ql_schema: Mapping[str, Any]
) -> None:
    _require(document.get("fixtureVersion") == "factory.interop-fixtures/v1", "unsupported fixture corpus version")
    validate_contract(document["contract"], schema, ql_schema)
    anti = document.get("antiFixtures")
    _require(isinstance(anti, list) and anti, "anti-fixture corpus is required")
    required_ids = {
        "action-as-capability-identity-collapse",
        "binding-as-ref",
        "model-as-agent",
        "session-as-agent",
        "provider-as-project",
        "stale-subject-state-evidence",
    }
    _require({item.get("id") for item in anti} == required_ids, "anti-fixture set is incomplete or contains unknown cases")
    for item in anti:
        _require(item.get("mustReject") is True, f"anti-fixture {item.get('id')} must be marked mustReject")
        try:
            reject_anti_fixture(item)
        except InteropError:
            continue
        raise InteropError(f"anti-fixture unexpectedly accepted: {item['id']}")


def round_trip(
    document: Mapping[str, Any], schema: Mapping[str, Any], ql_schema: Mapping[str, Any]
) -> dict[str, Any]:
    encoded = json.dumps(document, sort_keys=True, separators=(",", ":"))
    decoded = json.loads(encoded)
    validate_fixture_document(decoded, schema, ql_schema)
    return decoded


def load_default(root: Path) -> tuple[dict[str, Any], dict[str, Any], dict[str, Any]]:
    schema = json.loads((root / "contracts/factory/interop.schema.json").read_text(encoding="utf-8"))
    ql_schema = json.loads((root / "contracts/factory/ql-mef-composition.schema.json").read_text(encoding="utf-8"))
    document = json.loads((root / "contracts/factory/fixtures/interop-v1.json").read_text(encoding="utf-8"))
    return schema, ql_schema, document
